aws/src/test/java/org/apache/iceberg/aws/s3/S3OutputStreamTest.java

```diff
@@ -133,6 +134,53 @@ public void testAbortMultipart() {
     }
   }
 
+  @Test(expected = IllegalStateException.class)
+  public void testFailWriteIntAfterAsyncError() throws Exception {
+    doThrow(new RuntimeException()).when(s3mock).uploadPart((UploadPartRequest) any(), (RequestBody) any());
+    S3OutputStream stream = new S3OutputStream(s3mock, randomURI(), properties);
+    stream.write(randomData(10 * 1024 * 1024));
+    final long startTs = System.currentTimeMillis();
+    while (stream.getAsyncError() == null && System.currentTimeMillis() - startTs <= 1000) {
+      Thread.sleep(1);
+    }
+    Assert.assertNotNull(stream.getAsyncError());
+    try {
+      stream.write(1);
+    } finally {
+      stream.close();
+    }
+  }
+
+  @Test(expected = IllegalStateException.class)
+  public void testFailWriteArrayAfterAsyncError() throws Exception {
+    doThrow(new RuntimeException()).when(s3mock).uploadPart((UploadPartRequest) any(), (RequestBody) any());
+    S3OutputStream stream = new S3OutputStream(s3mock, randomURI(), properties);
+    stream.write(randomData(10 * 1024 * 1024));
+    final long startTs = System.currentTimeMillis();
+    while (stream.getAsyncError() == null && System.currentTimeMillis() - startTs <= 1000) {
+      Thread.sleep(1);
+    }
+    Assert.assertNotNull(stream.getAsyncError());
+    try {
+      stream.write(new byte[16]);
+    } finally {
+      stream.close();
+    }
+  }
+
+  @Test(expected = IllegalStateException.class)
+  public void testFailCloseAfterAsyncError() throws Exception {
+    doThrow(new RuntimeException()).when(s3mock).uploadPart((UploadPartRequest) any(), (RequestBody) any());
+    S3OutputStream stream = new S3OutputStream(s3mock, randomURI(), properties);
+    stream.write(randomData(10 * 1024 * 1024));
+    final long startTs = System.currentTimeMillis();
+    while (stream.getAsyncError() == null && System.currentTimeMillis() - startTs <= 1000) {
+      Thread.sleep(1);
+    }
+    Assert.assertNotNull(stream.getAsyncError());
+    stream.close();
```

Comment:
if you were being really strict you'd want to verify th multipart complete operation was never called

Reply:
good idea. added.